substreams-helper/src/storage.rs

```diff
@@ -1,31 +1,121 @@
+/// respective types.
+pub trait StorageLayout: Downcast {
+    /// Returns the size in bytes of this type. In the case of structs, this is the sum of all the slots
+    /// the struct takes, not only the raw addition of its underlying types.
+    fn size(&self) -> usize;
+    /// Will attempt to decode a set of slots into the given type. For types that can be offseted inside a slot,
+    /// there is an optional offset parameter. This will be ignored by arrays, mappings and structs.
+    fn decode(
+        &mut self,
+        slots: Vec<Vec<u8>>,
+        offset: Option<usize>,
+    ) -> Result<(), StorageDecodingError>;
+
+    /// Sets the slot for this type. This is necessary for when nesting different StorageLayout's, so that the parent can
+    /// calculate the slot for the child automatically (from its own slow).
+    fn set_slot(&mut self, slot: BigInt);
+}
+impl_downcast!(StorageLayout);
+
+pub trait ABIEncodeable {
+    fn abi_token(&self) -> ethabi::Token;
+    fn abi_decode(abi_encoded: Vec<u8>) -> Result<Self, ethabi::Error>
+    where
+        Self: Sized;
+}
+
+pub fn get_keccak_preimages_for_addresses(
+    store: &impl HasAddresser,
+    block: &pbeth::v2::Block,
+) -> Vec<Vec<u8>> {
+    let mut preimages: Vec<Vec<u8>> = vec![];
+    for call in block.calls() {
+        for change in &call.call.storage_changes {
+            if !store.has_address(Address::from_slice(change.address.as_slice())) {
+                continue;
+            }
+
+            preimages.extend(
+                call.call
+                    .keccak_preimages
+                    .values()
+                    .map(|preimage| -> Vec<u8> { substreams::Hex::decode(preimage).unwrap() }),
+            );
+            break;
+        }
+    }
+    preimages
+}
 
 /// Retrieves all storage changes in a block for all contracts under HasAddresser.
 pub fn get_storage_changes_for_addresses(
     store: &impl HasAddresser,
     block: &pbeth::v2::Block,
-) -> Vec<pbeth::v2::StorageChange> {
-    let mut changes: Vec<pbeth::v2::StorageChange> = vec![];
+) -> Vec<StorageChange> {
+    let mut changes: Vec<StorageChange> = vec![];
     for call in block.calls() {
         for change in &call.call.storage_changes {
             if !store.has_address(Address::from_slice(change.address.as_slice())) {
                 continue;
             }
-            changes.push(change.to_owned());
+
+            let mut preimage = None;
+            let preimage_bytes = call
+                .call
+                .keccak_preimages
+                .get(&substreams::Hex::encode(change.key.clone()));
+            if preimage_bytes.is_some() {
+                preimage = Some(substreams::Hex::decode(preimage_bytes.unwrap()).unwrap());
+            }
+            changes.push(StorageChange {
+                change: change.to_owned(),
+                preimage: preimage,
+            });
         }
     }
     changes
 }
 
+pub fn keccak256(data: Vec<u8>) -> Vec<u8> {
```

Comment:
Just thinking - would it be possible to set the signature as [0u8; 32] instead of Vec<u8> as it's more informative of the kekak data? Your abi_decode fn can take T:Into<Vec<u8>> instead or something similar to keep the logic the same if that works for you?

Reply:
Just checked and it looks like you could even just accept slices instead based on the current impls of the trait?

Reply:
Yup, just found it easier to work with vecs 😅 

Changed to slices